### DISCUSSION TITLE
On the problem of functional expression of porosity

### DISCUSSION THREAD

**Discussion by Devin-one** [PMEAL/OpenPNM#2369] [Q&A]
Dear professor:
    I'm currently working on gobs where permeability coefficient and porosity are related.  But this porosity is a distribution function.  Excuse me, how to express such a function in openpnm? In addition, I would like to ask how to express some function problems in openpnm? Sincerely look forward to your reply.

**Comment by jgostick** [comment]
Hi @Liqiwen123 
Maybe you should email me directly, and send few slides illustrating your idea, because I don't quite know how to answer you based on the above description.  I *think* you're asking about putting porosity gradients in the pore network, which is possible but there are several ways it could be done.

**Comment by Devin-one** [comment]


**Reply by jgostick** [reply]
I think you'd better edit that post!  It's all garbled.  Maybe you sent it via email and github messed up it?  If you edit in the textbox above, hit 'preview' to make sure it renders as you hoped.

**Comment by Devin-one** [comment]
Sorry, I'm really sorry. I sent it via email but github messed up it. To avoid errors again, I uploaded a file here. And I also have now resent an email with an attachment to your mailbox. Please check and guide. Thank you again.
[The previous e-mail.pdf](https://github.com/PMEAL/OpenPNM/files/8487448/The.previous.e-mail.pdf)

**Reply by jgostick** [reply]
Hi @Liqiwen123 
I'm sorry that I still don't know exactly what your try to accomplish.  The picture you include in the pdf suggest that K (darcy permeability) is a function of X and Y, so you want the pore network to have variable permeability.  This can be accomplished in a few ways:

1.  You can create a 'cubic lattice' network, where the pore size is a function of spatial location. Since the throat size is correlated with the pore size, and permeability is a strong function of pore/throat size, then the permeability will vary with location too.  You can do this as follows:
* Create a cubic network of say [60, 60, 60] nodes.
* Create a numpy array of the same size, with random values between 0 and 1 (``im1 = np.random.rand(60, 60, 60)``
* Create a second array, also the same size, that express the gradient you're looking for (``im2``)
* Multiply the two arrays and normalize so all values are between 0 and 1 (``normalized_im = im1*im2/(im1*im2).max()``.  These values will become the random seed values used by openpnm to generate pore sizes.  You put these into openpnm as ``pn['pore.seed'] = normalized_im.flatten()``
* You then tell openpnm which pore size distribution to use, like a weibull or whatever, and it will use ``'pore.seed'`` to generate these sizes, resulting in a spatially correlated pore and throat size distribution.
* You'll have to fiddle with the parameters of the distributions, and spacing of the network etc, to get thing working as desired.
2. You could create an image of a porous material with the structure you're looking for, then do an network extraction using ``porespy.networks.snow2``.  You could create such an image by adding spheres of different sizes to different locations or whatever you can come up with.  The extracted network will then have the pore size, porosity and permeability gradients that you imposed on the image.  

Both options are quite feasible.  Hope that helps.
